src/logics/magicsea/logic.swap-token.ts

```diff
@@ -0,0 +1,339 @@
+        const [token0, token1] = tokenA.sortsBefore(tokenB) ? [tokenA, tokenB] : [tokenB, tokenA];
+
+        const pair = new Pair(
+          new TokenAmount(token0, reserve0.toString()),
+          new TokenAmount(token1, reserve1.toString())
+        );
+
+        accumulator.push(pair);
+      }
+
+      return accumulator;
+    }, [] as Pair[]);
+
+    return pairs;
+  }
+
+  generateAllRoutePairs(tokenA?: Token, tokenB?: Token): [Token, Token][] {
+    const allBases = getBaseTokens(this.chainId).map((baseToken) => toUniToken(baseToken));
+
+    const basePairs: [Token, Token][] = [];
+    for (let i = 0; i < allBases.length; i++) {
+      for (let j = i + 1; j < allBases.length; j++) {
+        basePairs.push([allBases[i], allBases[j]]);
+      }
+    }
+
+    return [
+      // the direct pair
+      [tokenA, tokenB],
+      // token A against all bases
+      ...allBases.map((base): [Token | undefined, Token] => [tokenA, base]),
+      // token B against all bases
+      ...allBases.map((base): [Token | undefined, Token] => [tokenB, base]),
+      // each base against all bases
+      ...basePairs,
+    ]
+      .filter((tokens): tokens is [Token, Token] => Boolean(tokens[0] && tokens[1]))
+      .filter(([t0, t1]) => t0.address !== t1.address);
+  }
+
+  async getPairAddresses(pairTokens: Token[][]): Promise<Array<string | undefined>> {
+    const pairAddresses = [];
+    const multicallRecords = [];
+    const callsGetPair: common.Multicall3.CallStruct[] = [];
+    const iface = MagicSeaFactory__factory.createInterface();
+
+    // compose multi-calls
+    for (let i = 0; i < pairTokens.length; i++) {
+      const tokenA = pairTokens[i][0];
+      const tokenB = pairTokens[i][1];
+
+      if (tokenA && tokenB && !tokenA.equals(tokenB)) {
+        const tokenIn = toToken(tokenA);
+        const tokenOut = toToken(tokenB);
+        const [token0, token1] = tokenIn.sortsBefore(tokenOut) ? [tokenIn, tokenOut] : [tokenOut, tokenIn];
+
+        const callDataGetPair = iface.encodeFunctionData('getPair', [token0.address, token1.address]);
+        const factorAddress = getContractAddress(this.chainId, 'Factory');
+        callsGetPair.push({ target: factorAddress, callData: callDataGetPair });
+        multicallRecords.push(true);
+      } else {
+        multicallRecords.push(false);
+      }
+    }
+
+    const [resultsGetPair] = await Promise.all([this.multicall3.callStatic.tryAggregate(false, callsGetPair)]);
+
+    // decode multicall results
+    let j = 0;
+    for (let i = 0; i < multicallRecords.length; i++) {
+      if (multicallRecords[i]) {
+        const resultGetPair = resultsGetPair[j];
+        if (resultGetPair.success && resultGetPair.returnData !== '0x') {
+          const [pairAddress] = iface.decodeFunctionResult('getPair', resultGetPair.returnData);
+          if (pairAddress === constants.AddressZero) pairAddresses.push(undefined);
+          else {
+            pairAddresses.push(pairAddress);
+          }
```

Comment:
we can simplify this into one line
```suggestion
          pairAddresses.push(pairAddress === constants.AddressZero ? undefined : pairAddress);
```